CesiumRasterOverlays/src/RasterOverlayUtilities.cpp

```diff
@@ -277,6 +284,177 @@ RasterOverlayUtilities::createRasterOverlayTextureCoordinates(
+    const std::vector<CesiumGeometry::TriangleClipVertex>& clipResult);
+
+void addEdge(
+    EdgeIndices& edgeIndices,
+    double thresholdU,
+    double thresholdV,
+    bool keepAboveU,
+    bool keepAboveV,
+    bool invertV,
+    const AccessorView<glm::vec2>& uvs,
+    const std::vector<uint32_t>& clipVertexToIndices,
+    const std::vector<CesiumGeometry::TriangleClipVertex>& complements,
+    const std::vector<CesiumGeometry::TriangleClipVertex>& clipResult);
+
+void addSkirt(
+    std::vector<float>& output,
+    std::vector<uint32_t>& indices,
+    std::vector<FloatVertexAttribute>& attributes,
+    const std::vector<uint32_t>& edgeIndices,
+    const glm::dvec3& center,
+    double skirtHeight,
+    int64_t vertexSizeFloats,
+    int32_t positionAttributeIndex);
+
+void addSkirts(
+    std::vector<float>& output,
+    std::vector<uint32_t>& indices,
+    std::vector<FloatVertexAttribute>& attributes,
+    CesiumGeometry::UpsampledQuadtreeNode childID,
+    SkirtMeshMetadata& currentSkirt,
+    const SkirtMeshMetadata& parentSkirt,
+    EdgeIndices& edgeIndices,
+    int64_t vertexSizeFloats,
+    int32_t positionAttributeIndex,
+    bool hasInvertedVCoordinate);
+
+bool isWestChild(CesiumGeometry::UpsampledQuadtreeNode childID) noexcept {
+  return (childID.tileID.x % 2) == 0;
+}
+
+bool isSouthChild(CesiumGeometry::UpsampledQuadtreeNode childID) noexcept {
+  return (childID.tileID.y % 2) == 0;
+}
+
+void copyImages(const Model& parentModel, Model& result);
+void copyMetadataTables(const Model& parentModel, Model& result);
+} // namespace
+
+/*static*/ std::optional<Model>
+RasterOverlayUtilities::upsampleGltfForRasterOverlays(
+    const Model& parentModel,
+    UpsampledQuadtreeNode childID,
+    bool hasInvertedVCoordinate,
+    const std::string_view& textureCoordinateAttributeBaseName,
+    int32_t textureCoordinateIndex) {
+  CESIUM_TRACE("upsampleGltfForRasterOverlays");
+  Model result;
+
+  // Copy the entire parent model except for the buffers, bufferViews, and
+  // accessors, which we'll be rewriting.
+  result.animations = parentModel.animations;
+  result.materials = parentModel.materials;
+  result.meshes = parentModel.meshes;
+  result.nodes = parentModel.nodes;
+  result.textures = parentModel.textures;
+  result.images = parentModel.images;
+  result.skins = parentModel.skins;
+  result.samplers = parentModel.samplers;
+  result.cameras = parentModel.cameras;
+  result.scenes = parentModel.scenes;
+  result.scene = parentModel.scene;
+  result.extensionsUsed = parentModel.extensionsUsed;
+  result.extensionsRequired = parentModel.extensionsRequired;
+  result.asset = parentModel.asset;
+  result.extras = parentModel.extras;
+
+  // TODO: check if this is enough, not enough, or overkill
+  result.extensions = parentModel.extensions;
+  // result.extras_json_string = parentModel.extras_json_string;
+  // result.extensions_json_string = parentModel.extensions_json_string;
+
+  copyImages(parentModel, result);
```

Comment:
This added line was mentioned as a fix in the PR.

@kring , do you have any suspicion that this might cause a performance problem?

More a sanity check question. I didn't see problems in my testing. But in theory it's copying image data, where it didn't copy it before.

Reply:
Yeah it could cause a small performance hit in certain cases. For example, when draping a raster overlay over a textured tileset like Google Photorealistic 3D Tiles. Previously, the decoded images were copied when upsampling, because these are stored in `gltf.images[n].cesium`. That's still true. But the image buffers prior to decoding (i.e., the original JPG or whatever) were not copied, and now they are. Instead, the image would previously end up referring to an invalid bufferView.

Given that upsampling isn't particularly performance-sensitive, I think the correctness is worth the cost here. But a nice improvement across the board - not just for upsampling - would be to _remove_ bufferViews (and sections of buffers) that hold the compressed images after we decompress them, in order to save memory. It's outside the scope of this PR though, I think.